### PR TITLE
PERF: _form_blocks

### PR DESCRIPTION
**Pull request by jbrockmendel**
```
from asv_bench.benchmarks.frame_ctor import *

self = FromArrays()

self.setup()

%timeit self.time_frame_from_arrays_int()
2.28 ms ± 67 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  # <- master
1.82 ms ± 5.34 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)  # <- PR
```

Orthogonal to #43142, one more optimization after these two are in.

### CODE REVIEW

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/pull/43144/checks?check_run_id=3386832245 looks like a real error

**Comment by jbrockmendel** [comment]
Huh can't reproduce the failures locally, will have to dig into this

**Comment by jbrockmendel** [comment]
OK, looks like in numpy<1.21 comparing np.dtype object to EA dtype object raises instead of returning False, so the itertools.groupby is failing.  will figure out a workaround

**Comment by jbrockmendel** [comment]
updated with np<1.21-compatiblity + green

**Comment by jbrockmendel** [comment]
@jreback gentle ping; hoping to get this in before you take off

File: pandas/core/internals/managers.py [hunk 1]
```diff
@@ -1863,63 +1859,56 @@ def construction_error(
 # -----------------------------------------------------------------------
 
 
-def _form_blocks(arrays: list[ArrayLike], consolidate: bool) -> list[Block]:
-
-    items_dict: DefaultDict[str, list] = defaultdict(list)
-
-    for i, name_idx in enumerate(range(len(arrays))):
+def _grouping_func(tup):
```
Comment (jreback) [inline pandas/core/internals/managers.py:1862]: if u can type in a follow on

**Review by jreback — APPROVED** [review]

**Pull request merged**